doc/cascadia/Cascading-Default-Settings.md

```diff
@@ -0,0 +1,399 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2019-05-31
+last updated: 2019-06-13
+issue id: #754
+---
+
+# Cascading Default + User Settings
+
+## Abstract
+
+This spec outlines adding support for a cascading settings model. In this model,
+there are two settings files, instead of one.
+
+1. The default settings file
+2. The user's settings file
+
+The default settings file would be a static, read-only file shipped with the
+terminal. The user settings file would then contain all the user's chosen
+customizations to the settings. These two files would then be composed together
+when the app is launched, so that the runtime settings are the union of both the
+defaults and whatever modifications the user has chosen. This will enable the
+app to always use a default schema that it knows will be valid, and minimize the
+settings that the user needs to customize.
+
+Should the settings schema ever change, the defaults file will change, without
+needing to re-write the user's settings file.
+
+
+## Inspiration
+
+Largely inspired by the settings model that both VS Code (and Sublime Text) use.
+
+### Goal: Minimize Writing to `profiles.json`
+
+We want to write the user settings file, `profiles.json`, as little as possible.
+Each time we write the file, there's the possiblity that we've re-ordered the
+keys, as `jsoncpp` provides no ordering guarantee of the keys. This isn't great,
+as each write of the file will randomly re-order the file.
+
+One of our overarching goals with this change should be to re-write the user
+settings file as little as possible.
+
+## Solution Design
+
+The settings are now composed from two files: a "Default" settings file, and a
+"User" settings file.
+
+### Default Settings
+
+We'll have a static version of the "Default" file hardcoded within the
+application. We'll write the hardcoded version of this file out to an actual
+file on disk, `defaults.json`, however, that file will not actually be used. The
+defaults file will contain some sort of disclaimer like:
+
+```json
+// This is an auto-generated file. Place any modifications to your settings in "profiles.json"
+```
+
+This disclaimer will help identify that the file shouldn't be modified.
+
+The `defaults.json` file will be written out whenever we detect that it doesn't exist.
+
+**TODO**: We also need to write it out if it changes, though ideally without
+actually loading the file at first load. Perhaps we could start another
+thread/`fire_and_forget` once we're loaded to re-scan the file and re-write it
+if it has changed.
+
+Because the `defaults.json` file is hardcoded within our application, we can use

```

Comment:
We should consider making it just be a file in the package root instead of like a `const std::string_view`

Reply:
See, I wanted to avoid reading two files from the disk, and _completely remove_ the ability for the user to change it. 

Can we mark a file in our package as read-only?

Reply:
A file in our package is inherently read-only; it lives in `C:\Program Files\WindowsApps\[this entire folder has an ACL that says the user cannot edit it]\`

Reply:
derp yea I'll put it there